### DISCUSSION TITLE
Saving the central sulcus as a mesh

### DISCUSSION THREAD

**Discussion by pierpy** [brainvisa/brainvisa.github.io#50] [Q&A] [answered]
We would like to save the central sulcus identified with morphologist as a mesh both in the native and in the thalairach space. 
I tried using the following code 

```
from soma import aims
graph = aims.read(gfile)
trans_tal = aims.GraphManip.talairach(graph)
vs = graph['voxel_size']
bck_types = ['aims_ss', 'aims_bottom', 'aims_other']
point_cloud = []
for vertex in graph.vertices():
    if 'aims_Tmtktri' in vertex and vertex['label'] in ['S.C._left', 'S.C._right']:
        mesh = vertex['aims_Tmtktri']
        np.save('filename', mesh)`
```

but I get a 3d matrix in which the point don't seem to be the expected mesh.
Can some one help me in this task?

**Comment by denisri** [comment, marked as answer]
`np.save()` is not designed for meshes: numpy is an array manipulation library, not a mesh library. In other words meshes are not numpy objects and are not handled by it.
You can save the mesh using `aims.write(mesh, 'filename.gii')`.
The mesh itself has 3 arrays, a vertices array, a polygons array, and a normals array (which are C++ vectors but have bridges with numpy):
```
vertices = np.asarray(mesh.vertex())  # this is a Nx3 array.
poly = np.asarray(mesh.polygon())  # this is a Px3 array
```
polygons are triplets of indices in the vertices array (a triangle joins 3 points)
See: https://brainvisa.info/pyaims/sphinx/index.html
Denis

**Reply by pierpy** [reply]
I used aims.write(mesh, 'mesh.gii') 
But i can not open the saved .nii file and also in the Brain Visa data browser the mesh.nii file appears as in the attached pictures.
I can not understand wath is going wrong. May be the problem is in the segmentation process?
![IMG-4492](https://user-images.githubusercontent.com/7381399/120631482-3a60d700-c468-11eb-995f-9dfa1f15f54e.jpeg)
I simply started the Morphologist pipeline on a T1 imported RMN

**Comment by denisri** [comment]
`.nii` (**N**ifti) is an image file format.
`.gii` (**G**ifti) is a mesh and texture file format.
Meshes cannot be saved as `.nii`, thus the writer has switched to another format (here `.mesh`, our historical binary format).

**Reply by pierpy** [reply]
OK many thaks. So how can I save the mesh in format that I can use outside brain visa?

**Reply by denisri** [reply]
If you are using a format known outside of brainvisa, yes. `.mesh` is our brainvisa format (so it's not known outside, even if it has the most "efficient" storage structure, it's a binary format), but `.gii` is the recommended one (designed for neuroimaging), otherwise there are also `.ply`, `.obj` (`.obj` is 2 distinct formats actually: wavefront and MNI formats) etc. To make simple: use `.gii`.

**Comment by pierpy** [comment]
Thank !! This works!!
Actually, we want to save the file as STL format...do you know if there is
a tool to convert .gii to .stl?
Thank again!

Il giorno gio 3 giu 2021 alle ore 12:58 Denis Rivière <
***@***.***> ha scritto:

> If you are using a format known outside of brainvisa, yes. .mesh is our
> brainvisa format (so it's not known outside, even if it has the most
> "efficient" storage structure, it's a binary format), but .gii is the
> recommended one (designed for neuroimaging), otherwise there are also .ply,
> .obj (.obj is 2 distinct formats actually: wavefront and MNI formats)
> etc. To make simple: use .gii.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/brainvisa/brainvisa.github.io/discussions/50#discussioncomment-820195>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABYKDF3JFE2BE4FDDVSUATTTQ5N33ANCNFSM46AL4ECA>
> .
>


-- 

Ing. Pierpaolo Croce, PhD.
Institute for Advanced Biomedical Technologies - ITAB
and
Department of Neuroscience and Imaging
University of Chieti "G. D'Annunzio"
Via dei Vestini, 33
66013 Chieti, Italy

**Reply by denisri** [reply]
It you need to convert to STL, you should probably use a software like Blender, which is not designed for neuroimaging, thus doesn't know `.gii`. `.ply` is a better intermediate format, then. Or Wavefront (but `.obj` is an ambiguous extension, you'd have to specify the exact format when writing: `aims.write(mesh, 'mesh.obj', format='WAVEFRONT')`)

**Reply by Hboni** [reply]
If you are interested, I already used [Surf_ice](https://www.nitrc.org/projects/surfice/) which is a software for surface rendering, and support `.gii` and `.stl`. It is the possible to create `.stl` files using `.gii` files.
I use to load a gii file inside Suf_ice and save it with STL file format without issues. (In the menu "Advanced" > "Save mesh")

**Reply by pierpy** [reply]
@denisri thank you !
@Hboni great thank!

**Comment by wsnyder4** [comment]
Hi @pierpy and @denisri !
I wanted to follow up on the question about mapping meshes into Talairach space, as my current approach using some code pulled from a few other forum posts has my sulcus meshes mis-aligned still (see image attached, with the example of 5 subjects' central sulcus mesh after this code). 

<img width="1161" alt="Screen Shot 2021-10-26 at 4 33 56 PM" src="https://user-images.githubusercontent.com/59845848/138912533-0c937933-c617-4815-b9f7-d59cab2cc0fc.png">

I am using this code below:
```
graph = aims.read( "left_sulcal_graph.arg" )
trm = aims.GraphManip.talairach(graph)
meshes = {}
for v in graph.vertices():
    if v.has_key( 'label' ) and v.has_key( 'aims_Tmtktri' ):
        label = v[ 'label' ]
        mesh = v[ 'aims_Tmtktri' ]
        if meshes.has_key( label ):
            # concatenate with previous meshes with the same label
            aims.SurfaceManip.meshMerge( meshes[ label ], mesh )
        else:
            meshes[ label ] = mesh

for label, mesh in meshes.iteritems():
    aims.SurfaceManip.meshTransform( mesh, tr )
    aims.write( mesh, label + '_tal' + '.gii' )
```

From this transformation, the sulcus meshes I get out seem to not be aligned still. I noticed that the transformation matrix from `aims.GraphManip.talairach(graph)` has a different Tx Ty and Tz (translation vector) in its 4th column than the Tx Ty and Tz from `graph['Talairach_translation']` and am wondering if this is related to the translation being off (although replacing the 4th column's first three values from  `trm` with the Tx Ty and Tz  in `graph['Talairach_translation']` still gives sulci that seem mis-aligned). 
I am also wondering if I am missing some other operation with referentials.

Thank you in advance for the help!
Will

**Comment by denisri** [comment]
Your code seems OK to me, so maybe the transformations in the graphs are not correct ? Are the sulcal graphs aligned when loaded through brainvisa viewers ? If not then there is a problem with the alignment of your data (have you used the SPM normalization method, or the internal AimsMIRegister in Morphologist ? The latter is definitely less robust and less precise, it's only a replacement when you don't have SPM)

The `Talairach_translation` attribute in graphs is an old format for the transformation. The new one has a different shape but both are still written and supported in graphs. You should not have to bother about it. The figures are different because (if I remember) the translation was applied prior to the linear matrix, and not the contrary as it is more standard.

**Reply by wsnyder4** [reply]
The sulcal graphs are aligned when loaded into Anatomist, and when I run the same code but without the transformation (`aims.SurfaceManip.meshTransform( mesh, tr )`) the individual sulcal meshes are also still aligned, so it is interesting that they become misaligned after being given the same affine transformation. 

I preprocessed with FreeSurfer, and so the talairach_auto.xfm file input to the Import FreeSurfer to Morphologist process I suppose is the basis for my transformation. My goal is to be able to have the meshes in the Talairach space used by other software as well.

**Comment by denisri** [comment]
If you display graphs in Anatomist directly, no transformation is applied by default: each sulcal graph is displayed in its own native coordinates system, which is normally in a roughly same orientation, but without any registration. This is why I was speaking of brainvisa, which actually loads and uses registration transformations when loading objects in Anatomist. Another way to actually use the registration information using only Anatomist, is is to use the option "referential/load information from file header" in Anatomist (select all loaded graphs and use the popup menu on selected objects). Then you will actually see if the registration info is OK or not. If "registered" data are worse than natively unaligned ones, then there is a problem, for sure...

**Reply by wsnyder4** [reply]
Thank for you all this information -- this has helped me understand my problem more, but some interesting things still pop up. 
When I load sulcal graphs into BrainVISA or anatomist (with the talariach-AC-PC-Anatomist referential loaded), the sulci are aligned as expected, and when I instead load their referential from the header, they all have different referentials and are not aligned (because they are each in native space here I assume). This is shown below with the referential pins showed.



<img width="992" alt="Screen Shot 2021-11-08 at 5 08 58 PM" src="https://user-images.githubusercontent.com/59845848/140786724-053f7070-024e-4d90-a985-6867f187ab69.png">
However, I get opposite behavior when creating meshes or also accessing other graph attributes (e.g. gravity_center / refgravity_center). The meshes I transformed using the code in my earlier post in this thread end up not-aligned (left side in image below). Additionally, the related "refgravity_center" for several sulci shown below (computed as an average of refgravity_center for all elementary folds for a given sulcus, weighted by the number of points in each elementary fold (point_number)) are distributed very widely and seem similarly non-registered (notice the scale). When not transforming though, the meshes appear to align when all loaded into anatomist, and the related gravity_center (not the "ref" version) has data points tightly packed around where you would expect for registered sulci. I reproduced this gravity_center behavior also using coordinates from the meshes' gifti files and got the same plots. (for frame of reference in the plots, green = OCCIPITAL, orange = S.Or.)


<img width="897" alt="Screen Shot 2021-11-08 at 5 09 14 PM" src="https://user-images.githubusercontent.com/59845848/140786758-911f6630-cca1-432b-ab28-735900e92359.png">

The meshes and internal "ref" (v.s. non-"ref") values therefore all seem swapped to me, and I am a bit confused at how this could arise (I am using a standard brainvisa pipeline using imported FreeSurfer files via the import to brainvisa process). I am wondering if this means the talairach measures in output sulcal morphometry csv files I created should also be questioned too then.

Thank you for any input towards helping me understand here!
P.S. some fun art I made of 100 subjects' sulci below

Best,
Will



<img width="1039" alt="Screen Shot 2021-11-08 at 5 09 28 PM" src="https://user-images.githubusercontent.com/59845848/140786794-dc7f9128-0e74-4548-82f4-69ba797979a0.png">

**Comment by denisri** [comment]
> When I load sulcal graphs into BrainVISA or anatomist (with the talariach-AC-PC-Anatomist referential loaded), the sulci are aligned as expected, and when I instead load their referential from the header, they all have different referentials and are not aligned (because they are each in native space here I assume). This is shown below with the referential pins showed.

It should be the contrary in Anatomist: when referentials are loaded, then transforms to Talairach space apply (each data has its own referential, with a transform to Talairach ref), otherwise they are considered all in the same ref (which is normally not true) and no transformation applies. However if all T1 images in your dataset have been normalized prior to getting into freesurfer/morphoilogist, they happen to be registered without applying further transformations. Some datasets are pre-processed and normalized like this (the HCP cohort for instance).
Anyway the application of the transforms should not mess things up, so there is a problem there. If data are already normalized, they should all get an (almost) identical transform, that's it...

You can check a few things in brainvisa, such as displaying the AC/PC points by using the viewer (eye icon) on the .APC file ("commissure_coordinates" parameter in Morphologist). If some of them are not at the correct position, this is a clue...

**Reply by wsnyder4** [reply]
I am using the UKB, but it does seem to be a transformation issue. The talairach registrations in freesurfer seem to align correctly when I compare the transformed orig.mgz to the template talairach brain, but the AC-PC points are off in all of the subjects following the Freesurfer import grey/white matter segmentation step. I appreciate the help -- I will have to keep exploring!
